### PR TITLE
gnrc_ipv6_nib: consider all local interfaces when looking for address

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

If a node has two interfaces A with `2001:16b8:45b5:9af8:5884:3bff:fe4f:a903` and B with `2001:16b8:45b5:9afa:5884:3bff:fe4f:a902` and receives a neighbor solicitation on A for an address configured on interface B, answer the neighbor solicitation instead of bailing out with

> Target address `2001:16b8:45b5:9afa:5884:3bff:fe4f:a902` is not assigned to the local interface

### Issues/PRs references

see https://github.com/RIOT-OS/RIOT/pull/8823#discussion_r651768593

### CODE REVIEW

File: sys/net/gnrc/network_layer/ipv6/nib/nib.c [hunk 1]
```diff
@@ -967,7 +975,7 @@ static void _handle_nbr_sol(gnrc_netif_t *netif, const ipv6_hdr_t *ipv6,
     }
 #endif  /* CONFIG_GNRC_IPV6_NIB_SLAAC */
     if (ipv6_addr_is_unspecified(&ipv6->src)) {
-        gnrc_ndp_nbr_adv_send(&nbr_sol->tgt, netif, &ipv6->src, false, NULL);
+        gnrc_ndp_nbr_adv_send(&nbr_sol->tgt, tgt_netif, &ipv6->src, false, NULL);
```
Comment (miri64) [inline sys/net/gnrc/network_layer/ipv6/nib/nib.c:978]: From what I understand If the neighbor is solicited via `netif` corresponds interface A in your description and `tgt_netif` interface B. Shouldn't the neighbor advertisement go out the same interface (`netif`) the solicitation was received on?

File: sys/net/gnrc/network_layer/ipv6/nib/nib.c [hunk 2]
```diff
@@ -1013,10 +1021,10 @@ static void _handle_nbr_sol(gnrc_netif_t *netif, const ipv6_hdr_t *ipv6,
         reply_aro = _copy_and_handle_aro(netif, ipv6, nbr_sol, aro, sl2ao);
         /* check if target address is anycast */
         if (netif->ipv6.addrs_flags[tgt_idx] & GNRC_NETIF_IPV6_ADDRS_FLAGS_ANYCAST) {
-            _send_delayed_nbr_adv(netif, &nbr_sol->tgt, ipv6, reply_aro);
+            _send_delayed_nbr_adv(tgt_netif, &nbr_sol->tgt, ipv6, reply_aro);
```
Comment (miri64) [inline sys/net/gnrc/network_layer/ipv6/nib/nib.c:1024]: Same here

File: sys/net/gnrc/network_layer/ipv6/nib/nib.c [hunk 3]
```diff
@@ -1013,10 +1021,10 @@ static void _handle_nbr_sol(gnrc_netif_t *netif, const ipv6_hdr_t *ipv6,
...
         }
         else {
-            gnrc_ndp_nbr_adv_send(&nbr_sol->tgt, netif, &ipv6->src,
+            gnrc_ndp_nbr_adv_send(&nbr_sol->tgt, tgt_netif, &ipv6->src,
```
Comment (miri64) [inline sys/net/gnrc/network_layer/ipv6/nib/nib.c:1027]: And here.

**Comment by benpicco** [comment]
Hm this works because the source is a global address so the interface is ignored.
When I set the 'correct' interface [`gnrc_ndp_nbr_adv_send()`](https://github.com/RIOT-OS/RIOT/blob/master/sys/net/gnrc/network_layer/ndp/gnrc_ndp.c#L335) will bail out.

```
2021-06-18 17:50:05,264 # Iface  7  HWaddr: 5A:84:3B:4F:A9:03 
2021-06-18 17:50:05,266 #           L2-PDU:1500  MTU:1500  HL:64  RTR  
2021-06-18 17:50:05,267 #           Source address length: 6
2021-06-18 17:50:05,268 #           Link type: wired
2021-06-18 17:50:05,269 #           inet6 addr: fe80::5884:3bff:fe4f:a903  scope: link  VAL
2021-06-18 17:50:05,271 #           inet6 addr: 2001:16b8:45f9:73f8:5884:3bff:fe4f:a903  scope: global  VAL
2021-06-18 17:50:05,272 #           inet6 group: ff02::2
2021-06-18 17:50:05,273 #           inet6 group: ff02::1
2021-06-18 17:50:05,274 #           inet6 group: ff02::1:ff4f:a903
2021-06-18 17:50:05,274 #           
2021-06-18 17:50:05,275 #           Statistics for Layer 2
2021-06-18 17:50:05,276 #             RX packets 7  bytes 626
2021-06-18 17:50:05,277 #             TX packets 5 (Multicast: 5)  bytes 438
2021-06-18 17:50:05,278 #             TX succeeded 0 errors 0
2021-06-18 17:50:05,279 #           Statistics for IPv6
2021-06-18 17:50:05,279 #             RX packets 7  bytes 528
2021-06-18 17:50:05,280 #             TX packets 5 (Multicast: 5)  bytes 368
2021-06-18 17:50:05,280 #             TX succeeded 5 errors 0
2021-06-18 17:50:05,280 # 
2021-06-18 17:50:05,281 # Iface  6  HWaddr: 5A:84:3B:4F:A9:02 
2021-06-18 17:50:05,281 #           L2-PDU:1500  MTU:1500  HL:64  RTR  
2021-06-18 17:50:05,282 #           Source address length: 6
2021-06-18 17:50:05,282 #           Link type: wired
2021-06-18 17:50:05,283 #           inet6 addr: fe80::5884:3bff:fe4f:a902  scope: link  VAL
2021-06-18 17:50:05,284 #           inet6 addr: 2001:16b8:45f9:73fa:5884:3bff:fe4f:a902  scope: global  VAL
2021-06-18 17:50:05,284 #           inet6 group: ff02::2
2021-06-18 17:50:05,285 #           inet6 group: ff02::1
2021-06-18 17:50:05,285 #           inet6 group: ff02::1:ff4f:a902
2021-06-18 17:50:05,285 #           
2021-06-18 17:50:05,286 #           Statistics for Layer 2
2021-06-18 17:50:05,286 #             RX packets 5  bytes 805
2021-06-18 17:50:05,287 #             TX packets 13 (Multicast: 13)  bytes 1206
2021-06-18 17:50:05,287 #             TX succeeded 0 errors 0
2021-06-18 17:50:05,288 #           Statistics for IPv6
2021-06-18 17:50:05,288 #             RX packets 5  bytes 735
2021-06-18 17:50:05,289 #             TX packets 13 (Multicast: 13)  bytes 1024
2021-06-18 17:50:05,289 #             TX succeeded 13 errors 0
2021-06-18 17:50:05,289 # 
2021-06-18 17:50:22,014 # nib: Handle packet (icmpv6->type = 135)
2021-06-18 17:50:22,015 # nib: Received valid neighbor solicitation on 7:
2021-06-18 17:50:22,016 #      - Target address: 2001:16b8:45f9:73fa:5884:3bff:fe4f:a902
2021-06-18 17:50:22,016 #      - Source address: 2001:16b8:45f9:73f8:a4ed:29ff:feec:d4f3
2021-06-18 17:50:22,017 #      - Destination address: ff02::1:ff4f:a902
2021-06-18 17:50:22,018 # ndp: send neighbor advertisement (iface: 6, tgt: 2001:16b8:45f9:73fa:5884:3bff:fe4f:a902, dst: 2001:16b8:45f9:73f8:a4ed:29ff:feec:d4f3, supply_tl2a: 1)
2021-06-18 17:50:22,019 # ndp: building target link-layer address option (l2addr: 5A:84:3B:4F:A9:02)
2021-06-18 17:50:22,020 # ndp: building neighbor advertisement message
2021-06-18 17:50:22,021 # nib: get next hop link-layer address of 2001:16b8:45f9:73f8:a4ed:29ff:feec:d4f3%6
2021-06-18 17:50:22,022 # nib: 2001:16b8:45f9:73f8:a4ed:29ff:feec:d4f3 is on-link or in NC, start address resolution
2021-06-18 17:50:22,023 # nib: resolve address 2001:16b8:45f9:73f8:a4ed:29ff:feec:d4f3%7 from neighbor cache
```

File: sys/net/gnrc/network_layer/ipv6/nib/nib.c (hunk 1 above)
Comment (benpicco) [reply sys/net/gnrc/network_layer/ipv6/nib/nib.c]: Hm this works because the source is a global address so the interface is ignored for sending.
When I set the 'correct' interface [`gnrc_ndp_nbr_adv_send()`](https://github.com/RIOT-OS/RIOT/blob/master/sys/net/gnrc/network_layer/ndp/gnrc_ndp.c#L335) will bail out.

```
2021-06-18 17:50:05,264 # Iface  7  HWaddr: 5A:84:3B:4F:A9:03 
2021-06-18 17:50:05,266 #           L2-PDU:1500  MTU:1500  HL:64  RTR  
2021-06-18 17:50:05,267 #           Source address length: 6
2021-06-18 17:50:05,268 #           Link type: wired
2021-06-18 17:50:05,269 #           inet6 addr: fe80::5884:3bff:fe4f:a903  scope: link  VAL
2021-06-18 17:50:05,271 #           inet6 addr: 2001:16b8:45f9:73f8:5884:3bff:fe4f:a903  scope: global  VAL
2021-06-18 17:50:05,272 #           inet6 group: ff02::2
2021-06-18 17:50:05,273 #           inet6 group: ff02::1
2021-06-18 17:50:05,274 #           inet6 group: ff02::1:ff4f:a903
2021-06-18 17:50:05,274 #           
2021-06-18 17:50:05,275 #           Statistics for Layer 2
2021-06-18 17:50:05,276 #             RX packets 7  bytes 626
2021-06-18 17:50:05,277 #             TX packets 5 (Multicast: 5)  bytes 438
2021-06-18 17:50:05,278 #             TX succeeded 0 errors 0
2021-06-18 17:50:05,279 #           Statistics for IPv6
2021-06-18 17:50:05,279 #             RX packets 7  bytes 528
2021-06-18 17:50:05,280 #             TX packets 5 (Multicast: 5)  bytes 368
2021-06-18 17:50:05,280 #             TX succeeded 5 errors 0
2021-06-18 17:50:05,280 # 
2021-06-18 17:50:05,281 # Iface  6  HWaddr: 5A:84:3B:4F:A9:02 
2021-06-18 17:50:05,281 #           L2-PDU:1500  MTU:1500  HL:64  RTR  
2021-06-18 17:50:05,282 #           Source address length: 6
2021-06-18 17:50:05,282 #           Link type: wired
2021-06-18 17:50:05,283 #           inet6 addr: fe80::5884:3bff:fe4f:a902  scope: link  VAL
2021-06-18 17:50:05,284 #           inet6 addr: 2001:16b8:45f9:73fa:5884:3bff:fe4f:a902  scope: global  VAL
2021-06-18 17:50:05,284 #           inet6 group: ff02::2
2021-06-18 17:50:05,285 #           inet6 group: ff02::1
2021-06-18 17:50:05,285 #           inet6 group: ff02::1:ff4f:a902
2021-06-18 17:50:05,285 #           
2021-06-18 17:50:05,286 #           Statistics for Layer 2
2021-06-18 17:50:05,286 #             RX packets 5  bytes 805
2021-06-18 17:50:05,287 #             TX packets 13 (Multicast: 13)  bytes 1206
2021-06-18 17:50:05,287 #             TX succeeded 0 errors 0
2021-06-18 17:50:05,288 #           Statistics for IPv6
2021-06-18 17:50:05,288 #             RX packets 5  bytes 735
2021-06-18 17:50:05,289 #             TX packets 13 (Multicast: 13)  bytes 1024
2021-06-18 17:50:05,289 #             TX succeeded 13 errors 0
2021-06-18 17:50:05,289 # 
2021-06-18 17:50:22,014 # nib: Handle packet (icmpv6->type = 135)
2021-06-18 17:50:22,015 # nib: Received valid neighbor solicitation on 7:
2021-06-18 17:50:22,016 #      - Target address: 2001:16b8:45f9:73fa:5884:3bff:fe4f:a902
2021-06-18 17:50:22,016 #      - Source address: 2001:16b8:45f9:73f8:a4ed:29ff:feec:d4f3
2021-06-18 17:50:22,017 #      - Destination address: ff02::1:ff4f:a902
2021-06-18 17:50:22,018 # ndp: send neighbor advertisement (iface: 6, tgt: 2001:16b8:45f9:73fa:5884:3bff:fe4f:a902, dst: 2001:16b8:45f9:73f8:a4ed:29ff:feec:d4f3, supply_tl2a: 1)
2021-06-18 17:50:22,019 # ndp: building target link-layer address option (l2addr: 5A:84:3B:4F:A9:02)
2021-06-18 17:50:22,020 # ndp: building neighbor advertisement message
2021-06-18 17:50:22,021 # nib: get next hop link-layer address of 2001:16b8:45f9:73f8:a4ed:29ff:feec:d4f3%6
2021-06-18 17:50:22,022 # nib: 2001:16b8:45f9:73f8:a4ed:29ff:feec:d4f3 is on-link or in NC, start address resolution
2021-06-18 17:50:22,023 # nib: resolve address 2001:16b8:45f9:73f8:a4ed:29ff:feec:d4f3%7 from neighbor cache
```

But I agree that it's pretty hacky to rely on this - let's adapt `gnrc_ndp_nbr_adv_send()` instead.

File: sys/net/gnrc/network_layer/ndp/gnrc_ndp.c [hunk 4]
```diff
@@ -331,8 +331,10 @@ void gnrc_ndp_nbr_adv_send(const ipv6_addr_t *tgt, gnrc_netif_t *netif,
     gnrc_netif_acquire(netif);
     do {    /* XXX: hidden goto */
         int tgt_idx;
+        gnrc_netif_t *tgt_netif = gnrc_netif_get_by_ipv6_addr(tgt);
 
-        if ((tgt_idx = gnrc_netif_ipv6_addr_idx(netif, tgt)) < 0) {
+        if ((tgt_netif == NULL) ||
+            (tgt_idx = gnrc_netif_ipv6_addr_idx(tgt_netif, tgt)) < 0) {
```
Comment (miri64) [inline sys/net/gnrc/network_layer/ndp/gnrc_ndp.c:337]: This seems redundant first you get the interface by the address, then you check if the interface is assigned to that interface. Trying to cover race conditions should not be necessary due to the `gnrc_netif_acquire(netif)` above, so I propose to make the checking of `tgt_idx` an assert after the if failed:

```c
if (tgt_netif == NULL) {
    DEBUG("ndp: tgt not assigned to interface. Abort sending\n");
    break;
}
tgt_idx = gnrc_netif_ipv6_addr_idx(tgt_netif, tgt);
assert(tgt_idx >= 0);
```

**Review by miri64 — CHANGES_REQUESTED** [review]

File: sys/net/gnrc/network_layer/ndp/gnrc_ndp.c (hunk 4 above)
Comment (benpicco) [reply sys/net/gnrc/network_layer/ndp/gnrc_ndp.c]: applied

**Comment by MrKevinWeiss** [comment]
This would be good to have for the release, if @miri64 can help push this along, great, otherwise I would need some details on how to test (in layman's terms).

**Review by miri64 — APPROVED** [review]
My change requests were addressed and I trust @benpicco's testing here. There is also a fix now to pick the target address' interface to check its state when sending a neighbor advertisement. So I'm fine with merging this. There is at least one run of the weekly release tests before feature freeze, so _if_ this breaks something we will notice before that.

**Comment by benpicco** [comment]
tbh I still don't understand why this is necessary with #16557 in place, but in practice I still see nodes solicit for addresses that should be off-link / on a different interface of an on-link node.

**Comment by miri64** [comment]
With https://github.com/RIOT-OS/RIOT/pull/16557 I am not quite sure, if it might be possible for a route to be "on-link" and thus https://github.com/RIOT-OS/RIOT/pull/16557 assuming a false precondition. This still makes sense as a fix since the disparity between network interface and target address is broken.
